hw/top_earlgrey/data/top_earlgrey.hjson

```diff
@@ -996,77 +996,230 @@
+      { name: 'IOC4'                                              },
+      { name: 'IOC5'                                              },
+      { name: 'IOC6'                                              },
+      { name: 'IOC7'                                              },
+      { name: 'IOC8'             , type: 'BidirOd'                },
+      { name: 'IOC9'             , type: 'BidirOd'                },
+      { name: 'IOC10'            , type: 'BidirOd'                },
+      { name: 'IOC11'            , type: 'BidirOd'                },
+      // IOR
+      { name: 'IOR0'                                              },
+      { name: 'IOR1'                                              },
+      { name: 'IOR2'                                              },
+      { name: 'IOR3'                                              },
+      { name: 'IOR4'                                              },
+      { name: 'IOR5'                                              },
+      { name: 'IOR6'                                              },
+      { name: 'IOR7'                                              },
+      { name: 'IOR8'             , type: 'BidirOd'                },
+      { name: 'IOR9'             , type: 'BidirOd'                },
+      { name: 'IOR10'            , type: 'BidirOd'                },
+      { name: 'IOR11'            , type: 'BidirOd'                },
+      { name: 'IOR12'            , type: 'BidirOd'                },
+      { name: 'IOR13'            , type: 'BidirOd'                },
+    ]
+  }
 
-  // TODO: need to overhaul this datastructure.
   pinmux: {
 
-    // Total number of Multiplexed I/O
-    //  All the input/outputs from IPs are muxed in pinmux, and it has # of I/O
-    //  talks to the outside of top_earlgrey.
-    //  This field will be replaced to the length of PAD if padctrl is defined
-    num_mio: 44
-
-    // Dedicated IO modules. The in/out ports of the modules below are connected
-    //  to TOP IO port through PADS directly. It bypasses PINMUX multiplexers
-    dio_modules: [
-      { name: "spi_device", pad: ["ChC[0..5]"] },
-      { name: "spi_host0",  pad: ["ChB[0..5]"] },
-      { name: "usbdev",     pad: ["ChA[0..8]"] },
+    // This defines the signal to PAD mapping for the dedicated IOs.
+    // If the signals are not mapped to a specific pad, they either have to be
+    // assigned to 'muxed', which means the signal goes into the pinmux,
+    // or 'custom', which behaves similarly to dedicated IO, but basically means
+    // that the connection must be made by hand in the template.
+    // If a signal is not specified here, it will be tied off to zero or
+    // left unconnected.
+    // Note that just mentioning the module instance name acts as a wildcard
+    // and includes all available IOs of that module in this list.
+    signals: [
+      // SPI Host0
+      {name: 'spi_host0.sck',     pad: 'SPI_HOST_CLK'},
```

Comment:
I started out with a similar format and then combined some fields to make it less verbose.
But I see that this may have become too overloaded.

I've reworked this a bit along the lines of what you have suggested.